src/main/java/reactor/core/publisher/MonoFilterWhen.java

```diff
@@ -45,7 +51,269 @@
+					super.onError(error);
+					return;
+				}
+
+				if (p instanceof Callable) {
+					Boolean u;
+
+					try {
+						u = ((Callable<Boolean>) p).call();
+					}
+					catch (Throwable ex) {
+						Exceptions.throwIfFatal(ex);
+						//TODO not really needed as we don't expect to get there if the Mono source errors
+						Exceptions.addThrowable(ERROR, this, ex);
+						filterState = STATE_RESULT;
+						super.onError(error);
+						return;
+					}
+
+					if (u != null && u) {
+						actual.onNext(t);
+					}
+				}
+				else {
+					FilterWhenInner inner = new FilterWhenInner(this, !(p instanceof Mono));
+					if (ASYNC_FILTER.compareAndSet(this, null, inner)) {
+						filterState = STATE_RUNNING;
+						p.subscribe(inner);
+					}
+				}
+			}
+			//TODO drop?
+		}
+
+		@Override
+		public void onComplete() {
+			if (filterState == STATE_FRESH) {
+				//there was no value, we can complete empty
+				super.onComplete();
+			}
+			//otherwise just wait for the inner filter to apply, rather than complete too soon
+		}
+
+		@Override
+		public void onError(Throwable t) {
+			Exceptions.addThrowable(ERROR, this, t);
+			if (filterState != STATE_RUNNING) {
+				super.onError(t);
+			}
+		}
+
+		@Override
+		public Throwable getError() {
+			return error;
+		}
+
+		@Override
+		public void cancel() {
+			if (super.state != CANCELLED) {
+				super.cancel();
+				upstream.cancel();
+				cancelInner();
+			}
+		}
+
+		void cancelInner() {
+			FilterWhenInner a = asyncFilter;
+			if (a != INNER_CANCELLED) {
+				a = ASYNC_FILTER.getAndSet(this, INNER_CANCELLED);
+				if (a != null && a != INNER_CANCELLED) {
+					a.cancel();
+				}
+			}
+		}
+
+		void innerResult(Boolean item) {
+			filterState = STATE_RESULT;
+			if (item != null && item) {
+				//will reset the value with itself, but using parent's `value` saves a field
+				complete(value);
+			}
+			else {
+				super.onComplete();
+			}
+		}
+
+		void innerError(Throwable ex) {
+			Exceptions.addThrowable(ERROR, this, ex);
+			super.onError(error);
+		}
+
+		void innerComplete() {
+			if (filterState == STATE_RUNNING) {
+				Throwable e = error;
+				if (e == null) {
+					super.onComplete();
+				}
+				else {
+					super.onError(e);
+				}
+			}
+		}
+
+		@Override
+		public Object upstream() {
+			return upstream;
+		}
+
+		@Override
+		public Object scan(Attr key) {
+			switch (key) {
+				case PARENT:
+					return upstream;
+				case CANCELLED:
+					return super.state == CANCELLED;
+				case ERROR:
+					return error;
+				case BUFFERED:
+					return filterState == STATE_RUNNING ? 1 : 0;
+				case PREFETCH:
+					return 1;
+				default:
+					return super.scan(key);
+			}
+		}
+
+		@Override
+		public Stream<? extends Scannable> inners() {
+			FilterWhenInner c = asyncFilter;
+			return c == null ? Stream.empty() : Stream.of(c);
+		}
+	}
+
+	static final class FilterWhenInner implements InnerConsumer<Boolean> {
+
+		final MonoFilterWhenSubscriber<?> parent;
```

Comment:
could be mutualized with `FluxFilterWhen.FilterWhenInner` but we'd need a common interface for `FluxFilterWhenSubscriber` and `MonoFilterWhenSubscriber`.